Add multicast subcommand

yushoyamaguchi opened the pull request:
## Change
This change adds cilium multicast sub command to manage multicast groups and subscribers throughout the cluster

Following commands are added
```
# show list of multicast group
- cilium multicast list group

# show list of subscribers
- cilium multicast list subscriber --all
- cilium multicast list subscriber --group-ip <group-ip>

# make specified multicast group and add CiliumInternalIP of all cilium nodes to the group in each node
- cilium multicast add --group-ip <group-ip>

# Remove CiliumInternalIP of all cilium nodes from specified multicast group and delete the group in each node
- cilium multicast del --group-ip <group-ip>
```
This is follow up of https://github.com/cilium/cilium-cli/issues/2574

## Display example
![cli-result](https://github.com/cilium/cilium-cli/assets/50731798/e20b7bc4-5621-4b4e-a0c1-71cc146e00fe)

## Additional
As for testing, we will proceed with the implementation of the connectivity test for multicast as described in this issue.(https://github.com/cilium/cilium-cli/issues/2615)

## Reference
Current way of setup multicast : https://github.com/cilium/cilium/blob/main/Documentation/network/multicast.rst

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,285 @@
+				}
+			}
+		}
+		fmt.Fprintf(m.params.Writer, "Added %d subscribers to multicast group %s\n\n", cnt, m.params.MulticastGroupIP)
```
Adding pod name in this message would be helpful.

harsimran-pabla commented on cli/multicast.go:
```diff
@@ -0,0 +1,131 @@
+	}
+	cmd := &cobra.Command{
+		Use:   "add",
+		Short: "Add CiliumInternalIPs of all nodes to the specified multicast group as subscribers in every cilium-agent",
```
I'd refrain from mentioning CiliumInternalIPs in the description. It is an internal detail; a user should not care about it.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,285 @@
+		return err
+	}
+
+	//Create a map of ciliumInternalIPs of all nodes
```
Consider creating a helper method for this; the same pattern is used in the add and del methods.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,285 @@
+			if ip.IP != "" && ciliumPod.Name != podName {
+				cmd = []string{"cilium-dbg", "bpf", "multicast", "subscriber", "add", m.params.MulticastGroupIP, ip.IP}
+				_, err := m.execInCiliumPod(ctx, ciliumPod, cmd)
+				if err == nil {
```
If you call this CLI multiple times - you'll receive an 'already exists' type of error. If that is the case, check the error type. The 'already exists' error can be ignored ( idempotent call ), and for any other error type, we fail and return.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	for _, ciliumPod := range ciliumPods {
+		//Print cilium pod info
```
similar comment: print should be removed.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,285 @@
+		//If there are not specified multicast group, continue
+		cmd := []string{"cilium-dbg", "bpf", "multicast", "subscriber", "list", m.params.MulticastGroupIP}
+		output, err := m.execInCiliumPod(ctx, ciliumPod, cmd)
+		if err != nil && output == "" {
```
It might be better to check error type.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,285 @@
+			if ip.IP != "" && ciliumPod.Name != podName {
+				cmd = []string{"cilium-dbg", "bpf", "multicast", "subscriber", "del", m.params.MulticastGroupIP, ip.IP}
+				_, err := m.execInCiliumPod(ctx, ciliumPod, cmd)
+				if err == nil {
```
Similar comment: need more robust error handling.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,285 @@
+				}
+			}
+		}
+		fmt.Fprintf(m.params.Writer, "Deleted %d subscribers to multicast group %s\n\n", cnt, m.params.MulticastGroupIP)
```
```suggestion
		fmt.Fprintf(m.params.Writer, "Deleted %d subscribers from multicast group %s\n\n", cnt, m.params.MulticastGroupIP)
```

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,285 @@
+		}
+		fmt.Fprintf(m.params.Writer, "Deleted %d subscribers to multicast group %s\n\n", cnt, m.params.MulticastGroupIP)
+
+		//Delete the multicast group
```
Maybe all the subscriber cleanup is unnecessary; just deleting the group should be enough.

harsimran-pabla reviewed (changes requested):
Thank you for the changes; overall, it looks good. A couple of areas that need work are concurrency and error handling.

yushoyamaguchi commented:
@harsimran-pabla 
Thank you so much for reviewing.

I'm sorry, I have a question.
![github_err_msg](https://github.com/cilium/cilium-cli/assets/50731798/af4645a4-2a68-42b6-adca-f83920038006)
These warning messages come from GitHub actions.
In other subcommands, variable name `cmd` is used.  Can I ignore these warning messages, or should I change variable name to `_` ?

yushoyamaguchi commented on cli/multicast.go:
```diff
@@ -0,0 +1,131 @@
+func newCmdMulticastList() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "Show list of somthing about multicast",
```
I've fixed

asauber reviewed (commented):
I will save a detailed review for after the current round of changes.

harsimran-pabla commented:
Thanks @yushoyamaguchi  for addressing these comments.

> In other subcommands, variable name cmd is used. Can I ignore these warning messages, or should I change variable name to _ ?

It would be best if you fixed these comments ( these are static syntax checkers and have to make linters happy ). 

>Second, setting resolved tag for the place I fixed is my role or your role? (Currently, I only put 👍)
I'm sorry for ignorance about OSS development.

If you have addressed the comment; feel free to resolve it. 

> In addition, I replied to some your fix request comments which I cannot find the way to fix.

Please add a comment whichever comment is not clear, I will try my best to explain it further.

harsimran-pabla commented:
General commit comment. 

- Keep each commit about logical change instead of iterations on this PR. 
- Add more details about the change in the commit description.

asauber commented on multicast/multicast.go:
```diff
@@ -0,0 +1,354 @@
+			// List multicast subscribers
+			cmd := []string{"cilium-dbg", "bpf", "multicast", "subscriber", "list", target}
+			output, err := m.client.ExecInPod(ctx, pod.Namespace, pod.Name, defaults.AgentContainerName, cmd)
+			//ToDo:detect error type and handle properly
```
nit: If you are going to leave this `//ToDo: ` here, which is okay, it should be formatted as `// TODO `. Same for other occurrences of `ToDo`, also inline comments should start like `// <comment>` rather than `//<comment>`.

asauber commented on multicast/multicast.go:
```diff
@@ -0,0 +1,354 @@
+}
+
+// ListSubscriber lists multicast subscribers in every node for the specified multicast group or all multicast groups
+func (m *Multicast) ListSubscriber() error {
```
This should be named `ListSubscribers`

asauber commented on multicast/multicast.go:
```diff
@@ -0,0 +1,354 @@
+}
+
+// ListGroup lists multicast groups in every node
+func (m *Multicast) ListGroup() error {
```
This should be named `ListGroups`

asauber commented on multicast/multicast.go:
```diff
@@ -0,0 +1,354 @@
+		return err_ret
+	}
+
+	var wg2 sync.WaitGroup
```
I'd recommend splitting this into a new function, if not purely for the sake of readability and avoiding the pitfalls of `wg1, wg2..`

yushoyamaguchi commented:
@asauber 
Thank you for your reviewing.
I'll fix all improvement point, then tell you.

yushoyamaguchi commented:
@harsimran-pabla @asauber 

Thank you so much for supporting.
I've finished fixing all points.

I'm sorry for making git log complicated, but all commits are grouped by meaning.
Only minor changes are put in [this commit](https://github.com/cilium/cilium-cli/pull/2620/commits/66f6108173f2db10e89487793f53966e83c117f7) by force push.

asauber reviewed (changes requested):
This is shaping up nicely. Thank you for the updates.

Would you mind squashing all of the commits which modify your own code introduced in this PR?

The purpose of individual commits in a ready-to-merge PR should be to introduce new functionality, each with a detailed description. (For an example of a description, we have this recent commit https://github.com/cilium/cilium-cli/pull/2569/commits/0dea3f7de7c67775c29f452fdcdce897f89a9486)

You can feel free to break up the PR into "additive" commits if you wish, for example, one commit for each newly added command.

yushoyamaguchi commented:
@asauber 
I combined all the changes into one commit by `git squash` and explained the changes in detail in the commit message.
https://github.com/cilium/cilium-cli/pull/2620/commits/a923c64cb8f53b10b81d88a6abe1a7fc146230e9

cc @harsimran-pabla

asauber reviewed (commented):
Thanks, looks good to me.

In order to fix the failing tests (required for merge), can you try rebasing this branch on main?

yushoyamaguchi commented:
@asauber 
Thank you for your reviewing.
I 've fixed some errors from golang compiler and rebased this branch on main.

asauber reviewed (commented):
Looks like this needs one more round of lint fixes

yushoyamaguchi commented:
@asauber 
All checks have passed.
Thank you.

asauber commented on multicast/multicast.go:
```diff
@@ -0,0 +1,438 @@
+			}
+
+			var formattedOutput strings.Builder
+			formattedOutput.WriteString(fmt.Sprintf("%-*s\t", maxNodeNameLength+5, pod.Spec.NodeName))
```
For tabular output, we prefer to use the `tabwriter` package for alignment.

See https://github.com/cilium/cilium-cli/blob/main/bgp/routes.go#L257 for an example.

asauber commented on multicast/multicast.go:
```diff
@@ -0,0 +1,438 @@
+				}
+			}
+
+			//Add all ciliumInternalIPs of all nodes to the multicast group as subscribers
```
nit: another pass for `// ` (space after comment marker), would be appreciated

yushoyamaguchi commented:
@asauber 
Thank you for reviewing.
I've fixed all points.

In [here](https://github.com/cilium/cilium-cli/pull/2620/files#diff-abe2c01976e393eb320f6c1520249d16541f969e33771b03acf3516d6fa3008cR186), there is a comment that it depends on the cilium-dbg code.


Since it is impossible in principle to eliminate the dependence on cilium-dbg in this program, I thought it would be a good idea to put a comment in both cilium-dbg and cilium-dbg where they affect each other.

asauber reviewed (approved):
very nice, thanks for the updates

harsimran-pabla commented on cli/multicast.go:
```diff
@@ -0,0 +1,132 @@
+		Writer: os.Stdout,
+	}
+	cmd := &cobra.Command{
+		Use:   "del",
```
nit: del -> delete. `del` can be added as an alias.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,435 @@
+	errCh := make(chan error, len(ciliumPods))
+	wg.Add(len(ciliumPods))
+
+	w := tabwriter.NewWriter(m.params.Writer, minWidth, 0, padding, paddingChar, 0)
```
I have a common concern regarding all these commands. Output is in custom tabular format, which is not ideal as some users may want to consume it in other formats, like JSON, to automate some workflows. 

Could you consider separating data collection and data output in these methods? Fetch data from all the nodes, then output data based on the user's specified format.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,435 @@
+		go func(pod corev1.Pod) {
+			defer wg.Done()
+			// List multicast groups
+			cmd := []string{"cilium-dbg", "bpf", "multicast", "group", "list"}
```
Consider passing `-o json` to cilium-dbg command, it would make data parsing much easier ( use json unmarshaller ).

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,435 @@
+			// Use a mutex to ensure safe concurrent writes.
+			var mu sync.Mutex
+			mu.Lock()
+			fmt.Fprintf(w, "%s\t%s\n", pod.Spec.NodeName, formattedOutput.String())
```
Similar comment regarding usage of json data from cilium-dbg.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,435 @@
+	}
+	ciliumPods := ciliumPodsList.Items
+
+	// Create a map of ciliumInternalIPs of all nodes
```
Is this data collection required? I do not see it being used anywhere.

harsimran-pabla reviewed (changes requested):
Thanks, @yushoyamaguchi; change is coming along nicely. I have a couple of suggestions.

- Consider a use case where someone might want JSON format to display output. 
- When fetching data from cilium pods, use `-o json`; it makes it much easier to parse this information.

yushoyamaguchi commented:
@harsimran-pabla 

Thank you for reviewing.

> Consider a use case where someone might want JSON format to display output.
> When fetching data from cilium pods, use -o json; it makes it much easier to parse this information.

I have addressed both of these requirements.
The output format is not changed.

Because the amount of code changes is significant, I split them into separate commits for better visibility.
If you prefer to have them in one commit when merging, please say so.

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,553 @@
+	IsRemote bool   `json:"IsRemote"`
+}
+
+type GroupForSubscriber struct {
```
```suggestion
type GroupSubscriberData struct {
```

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,553 @@
+	Subscribers  []Subscriber `json:"subscribers"`
+}
+
+type NodeForSubscriber struct {
```
```suggestion
type NodeSubscriberData struct {
```

harsimran-pabla commented on multicast/multicast.go:
```diff
@@ -0,0 +1,553 @@
+	Groups []GroupForSubscriber `json:"groups"`
+}
+
+type NodeForGroup struct {
```
```suggestion
type NodeGroupData struct {
```

harsimran-pabla reviewed (changes requested):
Thanks @yushoyamaguchi, for bearing with me on this PR. There are a couple of things I would like to highlight: minor naming changes. And channel creations can be cleaned up. I'd suggest consolidating the channels, as shown in the example in one of the comments.

yushoyamaguchi commented:
@harsimran-pabla 

Thank you for kindly pointing out the places that needed improvement.
I'm happy because I could learn a lot.

I've fixed all the points you mentioned.

yushoyamaguchi commented:
@harsimran-pabla 
Thank you so much for the detailed review
I've squashed.

yushoyamaguchi commented:
@harsimran-pabla @asauber 
Thank you for the approval.
Should I do something else for merging?
As I said in [this PR](https://github.com/cilium/cilium-cli/pull/2724#issue-2430113723), I want to add test scenario using this multicast subcommand after merging...

-----
@fujitatomoya -san 's approval is required?
He said maybe not required but I can ask him if needed, because he is a my colleague.

harsimran-pabla commented:
Thanks, @yushoyamaguchi. I think this PR is ready to merge; I believe all concerns are addressed. We can move forward with e2e testing independently.

harsimran-pabla commented:
I'd suggest reaching out to @fujitatomoya to get his approval as well, since he reviewed this PR.

yushoyamaguchi commented:
Thank you.

fujitatomoya reviewed (approved):
@harsimran-pabla @asauber thanks for the reviews, im good to go.

yushoyamaguchi commented:
@harsimran-pabla @asauber @michi-covalent @fujitatomoya 
Thank you so much!
Should I write some Document?
